### PR TITLE
[PoC] Merge #2035 into 4.x

### PR DESCRIPTION
**Pull request by phansys**

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Do you need some help @phansys ?

**Comment by phansys** [comment]
>Do you need some help @phansys ?

Sure! I didn't check yet if the raised errors are caused by changes we must address in 4.x or if I must apply some changes on #2035.

File: src/Admin/BaseMediaAdmin.php [hunk 1]
```diff
@@ -37,6 +37,9 @@ abstract class BaseMediaAdmin extends AbstractAdmin
 
     protected ?ContextManagerInterface $contextManager = null;
 
+    /**
+     * @var string
```
Comment (VincentLanglet) [inline src/Admin/BaseMediaAdmin.php:41]: Introduced by https://github.com/sonata-project/SonataMediaBundle/pull/2035/files#diff-f7d08d0a17e3f0b38e23f90b6a73f036621412f2e2c74be531256496e1715ac8R45

best would be to remove this phpdoc to rely on the default one.

File: src/Model/GalleryManager.php [hunk 2]
```diff
@@ -13,6 +13,10 @@
 
 namespace Sonata\MediaBundle\Model;
 
+/**
+ * @phpstan-template T of GalleryInterface
+ * @phpstan-implements GalleryManagerInterface<T>
```
Comment (VincentLanglet) [inline src/Model/GalleryManager.php:18]: GalleryManagerInterface is not a generic in 4.x but was made as generic in 
https://github.com/sonata-project/SonataMediaBundle/pull/2035/files#diff-36a18f0b194491c1f64cdf74f503a7f8c4c7c0fd72b2ff3f08d6ab77b5e65dc2R21-R25

@jordisala1991 just did 
https://github.com/sonata-project/SonataMediaBundle/blame/4.x/src/Model/GalleryManagerInterface.php#L18-L20
and then there was no need for generic for every manager related classes.

File: src/Model/NoDriverGalleryManager.php [hunk 3]
```diff
@@ -19,6 +19,11 @@
  * @internal
  *
  * @author Andrey F. Mindubaev <covex.mobile@gmail.com>
+ *
+ * @phpstan-template T of GalleryInterface
+ * @phpstan-implements GalleryManagerInterface<T>
+ * @phpstan-template M of MediaInterface
+ * @phpstan-implements MediaManagerInterface<M>
```
Comment (VincentLanglet) [inline src/Model/NoDriverGalleryManager.php:26]: same

File: tests/Admin/TestMediaAdmin.php [hunk 4]
```diff
@@ -14,7 +14,12 @@
 namespace Sonata\MediaBundle\Tests\Admin;
 
 use Sonata\MediaBundle\Admin\BaseMediaAdmin;
+use Sonata\MediaBundle\Model\MediaInterface;
 
+/**
+ * @phpstan-template T of MediaInterface
+ * @phpstan-extends BaseMediaAdmin<T>
```
Comment (VincentLanglet) [inline tests/Admin/TestMediaAdmin.php:21]: same with https://github.com/sonata-project/SonataMediaBundle/blob/4.x/src/Admin/BaseMediaAdmin.php#L30

File: tests/App/Entity/Gallery.php [hunk 5]
```diff
@@ -15,10 +15,14 @@
...
 use Doctrine\ORM\Mapping as ORM;
 use Sonata\MediaBundle\Entity\BaseGallery;
+use Sonata\MediaBundle\Model\GalleryHasMediaInterface;
 
 /**
  * @ORM\Entity
  * @ORM\Table(name="media__gallery")
+ *
+ * @phpstan-template T of GalleryHasMediaInterface
+ * @phpstan-extends BaseGallery<T>
```
Comment (VincentLanglet) [inline tests/App/Entity/Gallery.php:25]: Gallery related classes/interface are not generic unlike https://github.com/sonata-project/SonataMediaBundle/pull/2035/files#diff-f77d00a1d76f9856d3b4df2488bcd5de873c644cf5d7aa153ca791493e6d1674R23

Main reason: methods `setGalleryHasMedias` and similar doesn't exist in 4.x https://github.com/sonata-project/SonataMediaBundle/blob/4.x/src/Model/GalleryInterface.php

File: tests/DependencyInjection/Compiler/ThumbnailCompilerPassTest.php [hunk 6]
```diff
@@ -24,6 +24,8 @@
 final class ThumbnailCompilerPassTest extends TestCase
 {
     /**
+     * @phpstan-param class-string $class
```
Comment (VincentLanglet) [inline tests/DependencyInjection/Compiler/ThumbnailCompilerPassTest.php:27]: duplicate of line 31

File: tests/Provider/DailyMotionProviderTest.php [hunk 7]
```diff
@@ -30,7 +30,9 @@
 use Sonata\MediaBundle\Thumbnail\FormatThumbnail;
 
 /**
- * @phpstan-extends AbstractProviderTest<\Sonata\MediaBundle\Provider\DailyMotionProvider>
+ * @phpstan-extends AbstractProviderTest<DailyMotionProvider>
+ *
+ * @property DailyMotionProvider $provider
```
Comment (VincentLanglet) [inline tests/Provider/DailyMotionProviderTest.php:35]: I don't think this property is true. We should just remove this line.

**Review by VincentLanglet — COMMENTED** [review]
All the errors are related to your changes @phansys.
It's mainly generics that I recommend to not add in 3.x because they don't exist in 4.x.

Or you should see with @jordisala1991 if it makes sens to introduce them.

**Comment by phansys** [comment]
>All the errors are related to your changes @phansys.

I know all the reported errors are introduced by the changes in #2035.
What I'm saying is that I don't analyzed yet if the offending changes must be reverted in that PR or if the required modifications must be propagated to 4.x.

**Comment by VincentLanglet** [comment]
> What I'm saying is that I don't analyzed yet if the offending changes must be reverted in that PR or if the required modifications must be propagated to 4.x.

Sure, that's why I listed them
- The generic-related must be reverted in the PR or added in 4.x, best would be to see with @jordisala1991 
- Other must be reverted in the PR

**Comment by core23** [comment]
Can we close this? 

Looks like a duplicate of #2215

**Comment by phansys** [comment]
>Can we close this?
>
>Looks like a duplicate of #2215

Not yet. It is not a duplicate. It is a PoC about the merge of #2035 into 4.x.

**Pull request closed without merging**